service/src/main/java/bio/terra/tanagra/db/AllArtifactsDao.java

```diff
@@ -0,0 +1,195 @@
+    Map<Integer, String> ogRevisionVerReviewIdMap = new HashMap<>(); // revisionVersion -> reviewId
+    Map<String, String> reviewIdMap = new HashMap<>(); // originalId -> clonedId
+    List<MapSqlParameterSource> reviewParamSets = new ArrayList<>();
+    List<MapSqlParameterSource> primaryEntityParamSets = new ArrayList<>();
+
+    ogReviews.forEach(
+        review -> {
+          String ogReviewId = review.getId();
+          String clonedId = ServiceUtils.newArtifactId();
+          ogRevisionIdReviewIdMap.put(review.getRevision().getId(), ogReviewId);
+          ogRevisionVerReviewIdMap.put(review.getRevision().getVersion(), ogReviewId);
+          reviewIdMap.put(ogReviewId, clonedId);
+
+          reviewParamSets.add(
+              reviewDao.buildReviewParam(
+                  clonedCohortId,
+                  clonedId,
+                  review.getSize(),
+                  review.getDisplayName(),
+                  review.getDescription(),
+                  userEmail));
+
+          Map<Literal, Integer> primaryEntityIdsToStableIndex =
+              reviewDao.getPrimaryEntityIdsToStableIndex(ogReviewId);
+          primaryEntityIdsToStableIndex.forEach(
+              (entityLiteral, index) ->
+                  primaryEntityParamSets.add(
+                      reviewDao.buildPrimaryEntityParam(
+                          clonedId, entityLiteral.getInt64Val(), index)));
+        });
+    reviewDao.insertReviewRows(reviewParamSets); // fk: cohort(id)
+    reviewDao.insertPrimaryEntityRows(primaryEntityParamSets); // fk: review(id)
+
+    // Revisions
+    Map<String, List<CohortRevision.CriteriaGroupSection>> clonedRevisionSections = new HashMap<>();
+    List<MapSqlParameterSource> revisionParamSets = new ArrayList<>();
+
+    ogCohort
+        .getRevisions()
+        .forEach(
+            revision -> {
+              String clonedId = ServiceUtils.newArtifactId();
+              clonedRevisionSections.put(clonedId, revision.getSections());
+              revisionParamSets.add(
+                  cohortDao.buildRevisionParam(
+                      clonedCohortId,
+                      clonedId,
+                      revision.getVersion(),
+                      revision.isMostRecent(),
+                      revision.isEditable(),
+                      userEmail,
+                      revision.getRecordsCount(),
+                      reviewIdMap.get(ogRevisionIdReviewIdMap.get(revision.getId()))));
+            });
+    String sql =
+        "INSERT INTO cohort_revision (cohort_id, id, review_id, version, is_most_recent, is_editable, created_by, last_modified_by, records_count) "
+            + "VALUES (:cohort_id, :id, :review_id, :version, :is_most_recent, :is_editable, :created_by, :last_modified_by, :records_count)";
+    cohortDao.insertRevisionRows(sql, revisionParamSets); // fk: cohort(id), review(id)
+
+    // Criteria, sections, groups, tags - fk: cohort_revision(id)
```

Comment:
ditto above: Can you do this copying in SQL instead of reading all the data into memory first? e.g. 
```
INSERT INTO criteria_group_section (cohort_revision_id, id, display_name, operator, is_excluded, ...)
SELECT :new_revision_id, id, display_name, operator, is_excluded, ...
FROM criteria_group_section
WHERE cohort_revision_id = :old_revision_id;
```

Reply:
done